Migrate compressor to a component

gh123man opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Migrate the compressor into a component keeping build tags local to each component implementation. 

This change is split up into 3 commits to ease reviewing:
Note: Each commit won't build on their own - it's mostly to separate the logic/strucutre changes from the very large number of imports/tests that needed to be updated. 

1. Migrate the compressor to a component (core logic) cfe2343de76f17fcf4187f6b75704541c644cac9
2. Update tests and imports ef550540ab4f8102ad8ccf5de154d7a332139d77
3. Add and update go modules 7ce0210952c8f5e9c5a61e5a266c8db211328693
### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

Change `serializer_compressor_kind` to each valid setting (`zlib` `zstd`), run the agent and ensure metrics are flowing. 

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

gh123man committed:
Migrate compressor to component

gh123man committed:
Update imports and tests

gh123man committed:
Update go mods

rayz commented on comp/serializer/compression/compressionimpl/zstd_and_zstd.go:
```diff
@@ -0,0 +1,43 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build zlib && zstd
```
File name should be `zlib_and_zstd`

gh123man replied:
oops - yeah

rayz commented on comp/README.md:
```diff
@@ -481,3 +481,9 @@ Package updater is the updater component.
+
+*Datadog Team*: agent-metrics-logs
+
+Package compression provides a compression implementation based on the configuration or avalible build tags.
```
```suggestion
Package compression provides a compression implementation based on the configuration or available build tags.
```

rayz commented on comp/serializer/compression/component.go:
```diff
@@ -1,34 +1,31 @@
 
-// Package utils provides a set of constants for compressing with zlib / zstd
-package utils
+// Package compression provides a compression implementation based on the configuration or avalible build tags.
```
nit: spelling `available`

rayz reviewed (commented):
minor nits

gh123man committed:
PR feedback

gh123man committed:
Remove stubs and remaining build tags

gh123man committed:
Merge branch 'main' into brian/AMLII-1544-compressor-component-2

# Conflicts:
#	go.sum
#	test/system/dogstatsd/receive_and_forward_test.go

dineshg13 commented on comp/serializer/compression/compressionimpl/strategy/zlib_strategy.go:
```diff
@@ -5,15 +5,15 @@
 
 //go:build zlib
```
Whats the need for this build tag ? We can remove this build tag rt ?

dineshg13 commented on comp/serializer/compression/compressionimpl/strategy/zstd_strategy.go:
```diff
@@ -5,15 +5,15 @@
 
 //go:build zstd
```
same comment as above

dineshg13 commented on comp/serializer/compression/compressionimpl/zlib_and_zstd.go:
```diff
@@ -0,0 +1,43 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build zlib && zstd
```
do we need this when we are reading from config ?

gh123man replied:
Yeah because there is a version of this without zstd. And in that case, the agent still respects the config but will not supply a zstd implementation. So we need to build tag them to preserve both build time compression options and runtime compression settings

dineshg13 commented on tasks/modules.py:
```diff
@@ -177,6 +177,10 @@ def dependency_path(self, agent_version):
     "comp/core/status": GoModule("comp/core/status", independent=True),
     "comp/core/status/statusimpl": GoModule("comp/core/status/statusimpl", independent=True),
+    "comp/serializer/compression": GoModule("comp/serializer/compression", independent=True),
+    "comp/serializer/compression/compressionimpl": GoModule(
+        "comp/serializer/compression/compressionimpl", independent=True
+    ),
```
Why do we need two separate modules ?

gh123man replied:
Removed

gh123man committed:
remove strategy build tags

gh123man committed:
Merge branch 'main' into brian/AMLII-1544-compressor-component-2

# Conflicts:
#	comp/otelcol/otlp/components/exporter/serializerexporter/go.mod
#	go.mod
#	go.sum
#	pkg/serializer/go.mod

gh123man committed:
missing file

gh123man committed:
remove compressionimpl mod

gh123man committed:
Fix mod errors

gh123man committed:
Merge branch 'main' into brian/AMLII-1544-compressor-component-2

GustavoCaso commented on comp/otelcol/otlp/components/exporter/serializerexporter/go.mod:
```diff
@@ -189,8 +191,8 @@ require (
 	go.uber.org/atomic v1.11.0 // indirect
-	go.uber.org/dig v1.17.0 // indirect
-	go.uber.org/fx v1.18.2 // indirect
+	go.uber.org/dig v1.17.1 // indirect
+	go.uber.org/fx v1.21.0 // indirect
```
I believe we should keep this dependencies version the same across all Go modules. Can we revert back this change?

gh123man replied:
I did a find replace and a full mod tidy - should be fixed now, thanks for pointing this out!

GustavoCaso commented on comp/serializer/compression/component.go:
```diff
@@ -1,34 +1,31 @@
 
-// Compressor is the interface for the compressor used by the Serializer
-type Compressor interface {
+// Component is the component type.
```
```suggestion
// Component is the component interface.
```

GustavoCaso replied:
Let's wait and do a batch update and modify the component template

GustavoCaso commented on comp/serializer/compression/go.mod:
```diff
@@ -0,0 +1,86 @@
+	github.com/DataDog/datadog-agent/pkg/util/fxutil v0.53.0-rc.2
+	github.com/DataDog/datadog-agent/pkg/util/log v0.53.0-rc.2
+	github.com/DataDog/zstd v1.5.5
+	go.uber.org/fx v1.21.0
```
Same comment about fx version as before https://github.com/DataDog/datadog-agent/pull/24012/files#r1546529123

GustavoCaso commented on go.mod:
```diff
@@ -263,8 +264,8 @@ require (
-	go.uber.org/dig v1.17.0
-	go.uber.org/fx v1.18.2
+	go.uber.org/dig v1.17.1
+	go.uber.org/fx v1.21.0
```
Sam as https://github.com/DataDog/datadog-agent/pull/24012/files#r1546529123

gh123man committed:
Go uber Fx -> v1.18.2

GustavoCaso commented on pkg/serializer/go.mod:
```diff
@@ -157,8 +159,8 @@ require (
-	go.uber.org/dig v1.17.0 // indirect
-	go.uber.org/fx v1.18.2 // indirect
+	go.uber.org/dig v1.17.1 // indirect
+	go.uber.org/fx v1.21.0 // indirect
```
Same as https://github.com/DataDog/datadog-agent/pull/24012/files#r1546529123

gh123man commented on comp/serializer/compression/compressionimpl/common.go:
```diff
@@ -0,0 +1,16 @@
+// Package compressionimpl provides a set of functions for compressing with zlib / zstd
+package compressionimpl
+
+// ZlibKind defines a const value for the zlib compressor
+const ZlibKind = "zlib"
+
+// ZstdKind  defines a const value for the zstd compressor
+const ZstdKind = "zstd"
+
+// NoneKind defines a const value for disabling compression
+const NoneKind = "none"
```
Today these are private to the implementation, and by design the interface doesn't care about what the underlying implementation ends up being (from build tags, or config). Perhaps this could change in the future - but this is why I put them in the impl today. 

With this in mind, do you still think I should move them to the interface?

GustavoCaso replied:
Thanks for the context. In that case it doesn't make sense to move to the component interface package

gh123man committed:
merge main into brian/AMLII-1544-compressor-component-2

# Conflicts:
#	cmd/cluster-agent-cloudfoundry/subcommands/run/command.go
#	cmd/cluster-agent/subcommands/start/command.go
#	comp/dogstatsd/server/server_test.go

gh123man committed:
fix go.uber.org/dig v1.17.0

GustavoCaso reviewed (approved):
LGTM 🎉

paulcacheux commented on comp/serializer/compression/go.mod:
```diff
@@ -0,0 +1,86 @@
+
+go 1.21.8
+
+replace (
```
Errors found:
  - github.com/DataDog/datadog-agent/comp/serializer/compression/go.mod is missing a replace for github.com/DataDog/datadog-agent/cmd/agent/common/path
  - github.com/DataDog/datadog-agent/comp/serializer/compression/go.mod is missing a replace for github.com/DataDog/datadog-agent/comp/core/flare/types
  - github.com/DataDog/datadog-agent/comp/serializer/compression/go.mod is missing a replace for github.com/DataDog/datadog-agent/comp/core/telemetry
  - github.com/DataDog/datadog-agent/comp/serializer/compression/go.mod is missing a replace for github.com/DataDog/datadog-agent/pkg/telemetry
  - github.com/DataDog/datadog-agent/comp/serializer/compression/go.mod is missing a replace for github.com/DataDog/datadog-agent/pkg/util/testutil

NouemanKHAL reviewed (approved):
Approved for `cmd/cluster-agent-cloudfoundry/subcommands/run/command.go`

gh123man committed:
Fix missing mod replace

KevinFairise2 reviewed (approved):
Looks good for the file owned by ADT

chouetz reviewed (approved):
ok for files owned by agent-ci-experience

duncanpharvey reviewed (changes requested):
Adding my review here so it's documented. Let's make sure the metric Serverless Integration tests are passing. The tests appear to begin failing on this commit: https://github.com/DataDog/datadog-agent/pull/24012/commits/11bc2d5594cee33f02663e0e351a7ac62a3de45c

gh123man committed:
Merge branch 'main' into brian/AMLII-1544-compressor-component-2

# Conflicts:
#	.gitlab/pkg_metrics/pkg_metrics.yml
#	comp/agent/bundle_test.go
#	pkg/aggregator/demultiplexer_test.go
#	pkg/collector/corechecks/oracle/testutil.go

gh123man committed:
Fix serverless integration tests

gh123man committed:
Fix tests

duncanpharvey commented:
Looks good from Serverless 👍

gh123man committed:
Merge branch 'main' into brian/AMLII-1544-compressor-component-2

# Conflicts:
#	go.mod
